pkg/pss/trojan.go

```diff
@@ -203,62 +201,54 @@ func contains(col Targets, elem []byte) bool {
 }
 
 // mine iteratively enumerates different nonces until the address (BMT hash) of the chunkhas one of the targets as its prefix
-func mine(ctx context.Context, odd bool, f func(nonce []byte) (swarm.Chunk, error)) (swarm.Chunk, error) {
+func mine(ctx context.Context, odd bool, f func(nonce []byte) (swarm.Chunk, error)) (r swarm.Chunk, err error) {
 	seeds := make([]uint32, 8)
-	for i := range seeds {
-		b, err := random.Int(random.Reader, maxUint32)
-		if err != nil {
-			return nil, err
-		}
-		seeds[i] = uint32(b.Int64())
-	}
 	initnonce := make([]byte, 32)
-	for i := 0; i < 8; i++ {
-		binary.LittleEndian.PutUint32(initnonce[i*4:i*4+4], seeds[i])
+	if _, err := io.ReadFull(random.Reader, initnonce); err != nil {
+		return nil, err
 	}
 	if odd {
 		initnonce[28] |= 0x01
 	} else {
 		initnonce[28] &= 0xfe
 	}
-	seeds[7] = binary.LittleEndian.Uint32(initnonce[28:32])
+	for i := range seeds {
+		seeds[i] = binary.BigEndian.Uint32(initnonce[i*4 : i*4+4])
+	}
 
-	quit := make(chan struct{})
-	// make both  errs  and result channels buffered so they never block
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	eg, ctx := errgroup.WithContext(ctx)
 	result := make(chan swarm.Chunk, 8)
-	errs := make(chan error, 8)
 	for i := 0; i < 8; i++ {
-		go func(j int) {
+		j := i
+		eg.Go(func() error {
 			nonce := make([]byte, 32)
 			copy(nonce, initnonce)
-			for seed := seeds[j]; ; seed++ {
-				binary.LittleEndian.PutUint32(nonce[j*4:j*4+4], seed)
+			for seed := seeds[j]; ; seed += 2 {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				default:
+				}
+				binary.BigEndian.PutUint32(nonce[j*4:j*4+4], seed)
 				res, err := f(nonce)
 				if err != nil {
-					errs <- err
-					return
+					return err
 				}
 				if res != nil {
 					result <- res
-					return
-				}
-				select {
-				case <-quit:
-					return
-				default:
+					return nil
 				}
 			}
-		}(i)
-	}
-	defer close(quit)
-	select {
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	case err := <-errs:
-		return nil, err
-	case res := <-result:
-		return res, nil
+		})
 	}
+	go func() {
+		err = eg.Wait()
+		result <- nil
+	}()
+	// if there was an error it gets set
+	return <-result, err
```

Comment:
I am still not 100% sure if we cover all the cases here. I would feel more comfortable if we could test some of these things in the unit tests. For eg. delaying some of the function calls passed as arguments here.